lib/Dataflow_analysis.ml

```diff
@@ -0,0 +1,800 @@
+  let data_vars = ExprSet.of_list [Var "x"] in
+  let prior_term_labels =
+    exprset_independent_target_terms df_graph data_vars
+  in
+  let prior_terms =
+    List.map (LabelSet.to_list prior_term_labels) ~f:(fun l ->
+        match (LabelMap.find_exn df_graph.node_info_map l).loc with
+        | TargetTerm {term; _} -> term
+        | _ -> raise (Failure "Found non-target term in target term list") )
+  in
+  let preexisting_vars =
+    ExprSet.of_list
+      (List.map ("target" :: Map.Poly.keys var_table) ~f:(fun v -> Var v))
+  in
+  if false then (
+    Sexp.pp_hum Format.std_formatter
+      [%sexp (df_graph.node_info_map : node_info_fixpoint LabelMap.t)] ;
+    print_string "\n\n" ;
+    print_endline
+      ( "Preexisting variables: "
+      ^ Sexp.to_string [%sexp (preexisting_vars : ExprSet.t)] ) ;
+    print_endline
+      ( "Target term nodes: "
+      ^ Sexp.to_string [%sexp (df_graph.target_term_nodes : LabelSet.t)] ) ;
+    print_endline
+      ( "Possible endpoints: "
+      ^ Sexp.to_string [%sexp (df_graph.possible_exits : LabelSet.t)] ) ;
+    print_endline
+      ( "Assumed data variables: "
+      ^ Sexp.to_string [%sexp (data_vars : ExprSet.t)] ) ;
+    print_endline
+      ( "Data-independent target term expressions: "
+      ^ Sexp.to_string [%sexp (prior_terms : expr list)] ) ;
+    print_endline
+      ( "Var " ^ var ^ " depends on labels: "
+      ^ Sexp.to_string [%sexp (label_deps : LabelSet.t)] ) ;
+    print_endline
+      ( "Var " ^ var ^ " depends on preexisting variables: "
+      ^ Sexp.to_string [%sexp (expr_deps : ExprSet.t)] ) ) ;
+  df_graph
+
+(***********************************)
+(* Tests                           *)
+(***********************************)
+
+(* Inline tests don't seem to work *)
+
+(* let%test _ = 5 = 120
+   let%test _ = raise (Failure "ran test")*)
+
+let%expect_test "Example program" =
+  let ast =
+    Parse.parse_string Parser.Incremental.program
+      "      model {\n\
+      \              for (i in 1:2)\n\
+      \                for (j in 3:4)\n\
+      \                  print(\"Badger\");\n\
+      \            }\n\
+      \            "
+  in
+  let mir =
+    Ast_to_Mir.trans_prog "" (Semantic_check.semantic_check_program ast)
+  in
+  let table, block = mir.modelb in
+  let df_graph = block_dataflow_graph block table in
+  print_s [%sexp (df_graph : dataflow_graph)] ;
+  [%expect
+    {|
+      ((node_info_map
+        ((0
+          ((dep_sets (() (((Var target) 0)))) (possible_previous ()) (rhs_set ())
+           (controlflow ()) (loc StartOfBlock)))
+         (1
+          ((dep_sets ((((Var target) 0)) (((Var i) 1) ((Var target) 0))))
+           (possible_previous (0)) (rhs_set ()) (controlflow (0))
+           (loc (MirNode "\"string\", line 2-4"))))
+         (2
+          ((dep_sets
+            ((((Var i) 1) ((Var j) 2) ((Var target) 0))
+             (((Var j) 2) ((Var target) 0))))
+           (possible_previous (1 2)) (rhs_set ()) (controlflow (1))
```

Comment:
You're right, I was really just using this to try to figure out expectation tests. I should try applying your fix so I can write some more interesting tests.

Yes, that's a bug because I haven't handled NRFunApp yet. It should at least be a node with a RHS, but ideally I need to actually traverse function bodies because (if I understand correctly) some functions can increment target.

Reply:
Yep. I was originally thinking we'd want to decorate statements with the Vars they affect (including FunDefs and `target`), though we'd probably need something like SSA to be able to be able to tell different values assigned to Vars apart.